### ISSUE TITLE
feature: pre-stored labels for multisig transactions

### ISSUE THREAD

**Issue by emergent-reasons** [Electron-Cash/Electron-Cash#1888]
When initiating a stack of multisig transactions, it seems the label is completely ignored.
The UX for then becomes that the user has to wait for an undefined period of time for the tx to be signed and broadcast. Then when the tx eventually appears in the history, it has no label. The outputs need to be verified against an externally recorded record to figure out which entry in the history belongs to which intended transaction. Then the label can be applied after the fact.

For even the small number of tx that I do this is a error-prone and time consuming process. It would be nice if

1. short term: multisig initiations that require 1+ additional sigs have the description field disabled (i.e. the last signer can fill out their own description). Leaving it enabled gives the false impression that the description will be recorded.
2. better: somehow ec records the description without having the txid yet available and is able to identify the tx when it appears. I assume this is very non-trivial, but at least I want to propose it.

I believe transaction labels are stored by txid and then appear when electron can confirm the presence of the txid on the network? I was thinking for this process:

1. Store the coins and outputs
2. Any time one of the coins is spent, it permanently pops the multisig from the record
3. If the spent coin happens to involve exactly the same inputs and outputs, then the recorded label is applied to the txid.
4. ??? No idea if that's a shit process or not. Just trying to leave a potential solution instead of only a problem.

**Comment by emergent-reasons** [comment]
I believe transaction labels are stored by txid and then appear when electron can confirm the presence of the txid on the network? I was thinking for this process:

Store the coins and outputs
Any time one of the coins is spent, it permanently pops the multisig from the record
If the spent coin happens to involve exactly the same inputs and outputs, then the recorded label is applied to the txid.
??? No idea if that's a shit process or not. Just trying to leave a potential solution instead of only a problem.

**Comment by emergent-reasons** [comment]
FWIW this is not high priority for me. Just want to capture the issue.

**Comment by cculianu** [comment]
Yes you are right @emergent-reasons -- labels are stored by txid and so a new "pre-signed" labeling scheme would have to be devised, which is keyed off something other than txid.

I envision it working something like this:

- Use an alternate key to store the "pre-signed labels". 
- This could be done by taking a hash of all the inputs and outputs for example
- Some time later when a tx comes in off the network, if the txid has no label, scan this set of "pre-signed labels" as a fall-back and see if maybe there is a label there
- If so, apply the label to the tx, deleting the pre-signed label to save space
- As a bonus these pre-signed labels should expire after some time to save on wallet disk space
